syntax = "proto3";

package rentflow;

// Сервис для управления арендой
service LeaseService {
  rpc CreateLease(CreateLeaseRequest) returns (CreateLeaseResponse);
  rpc GetLease(GetLeaseRequest) returns (GetLeaseResponse);
  
  // --- ИЗМЕНЕНО ---
  // Убедимся, что UpdateLease определен
  rpc UpdateLease(UpdateLeaseRequest) returns (UpdateLeaseResponse);
  // Оставим ListLeases, он не мешает
  rpc ListLeases(ListLeasesRequest) returns (ListLeasesResponse);
  
  // --- ДОБАВЛЕНО ---
  // Добавляем метод TerminateLease
  rpc TerminateLease(TerminateLeaseRequest) returns (TerminateLeaseResponse);
  // --- КОНЕЦ ДОБАВЛЕНИЯ ---
}

// Модели (Без изменений)
message Tenant {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string phone = 5;
}

message Property {
  int32 id = 1;
  string address = 2;
  string city = 3;
  string country = 4;
}

message Lease {
  int32 id = 1;
  Property property = 2;
  Tenant tenant = 3;
  string start_date = 4;      // YYYY-MM-DD
  string end_date = 5;        // YYYY-MM-DD
  float monthly_rent = 6;
  int32 payment_day = 7;      // день месяца (1-31)
}

// Запросы и ответы CreateLease (Без изменений)
message CreateLeaseRequest {
  Property property = 1;
  Tenant tenant = 2;
  string start_date = 3;
  string end_date = 4;
  float monthly_rent = 5;
  int32 django_lease_id = 6;
  int32 payment_day = 7;
}

message CreateLeaseResponse {
  bool success = 1;
  Lease lease = 2;
  string message = 3;
}

// GetLease (Без изменений)
message GetLeaseRequest {
  int32 lease_id = 1;
}

message GetLeaseResponse {
  bool found = 1;
  Lease lease = 2;
}

// --- ИЗМЕНЕНО (UPDATE) ---
// Это сообщение вызывает ошибку 'AttributeError', так как его не было.
message LeaseUpdate {
  optional float monthly_rent = 1;
  optional string end_date = 2;
}

message UpdateLeaseRequest {
  // Используем int32, чтобы соответствовать типу 'id' в Lease
  int32 fastapi_lease_id = 1; 
  int32 django_lease_id = 2;  // ID из Django (для Kafka)
  LeaseUpdate update_data = 3;  // Вложенное сообщение с обновлениями
}

message UpdateLeaseResponse {
  bool success = 1;
  Lease lease = 2; // Оставляем, как у вас было
}
// --- КОНЕЦ ИЗМЕНЕНИЯ (UPDATE) ---


// ListLeases (Без изменений)
message ListLeasesRequest {
  int32 page = 1;
  int32 page_size = 2;
}

message ListLeasesResponse {
  repeated Lease leases = 1;
  int32 total_count = 2;
}


// --- ДОБАВЛЕНО (TERMINATE) ---
message TerminateLeaseRequest {
  int32 fastapi_lease_id = 1;
  int32 django_lease_id = 2;
  string actual_end_date = 3; // YYYY-MM-DD
}

message TerminateLeaseResponse {
  bool success = 1;
  string message = 2;
}
// --- КОНЕЦ ДОБАВЛЕНИЯ (TERMINATE) ---